modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/FullStateTransferIndexChooser.java

```diff
@@ -273,137 +261,37 @@ private void onIndexRemoved(RemoveIndexEventParameters parameters) {
-
-                    if (index.status() == AVAILABLE) {
-                        // On drop table event.
-                        readOnlyIndexes.add(new ReadOnlyIndexInfo(index, catalogActivationTimestampBusy(catalogVersion), catalogVersion));
-                    } else if (index.status() == STOPPING) {
-                        readOnlyIndexes.add(
-                                new ReadOnlyIndexInfo(index, findStoppingActivationTsBusy(indexId, catalogVersion - 1), catalogVersion)
-                        );
-                    } else {
-                        // Index that is dropped before even becoming available.
-                        tableVersionByIndexId.remove(indexId);
-                    }
+                    return;
                 }
-            });
-        });
-    }
 
-    private void onIndexCreated(CreateIndexEventParameters parameters) {
-        inBusyLock(busyLock, () -> {
-            CatalogIndexDescriptor index = parameters.indexDescriptor();
-
-            int tableVersion = tableVersionBusy(index, parameters.catalogVersion());
+                IndexMeta indexMeta = indexMetaStorage.indexMeta(indexId);
 
-            tableVersionByIndexId.put(index.id(), tableVersion);
-        });
-    }
+                assert indexMeta != null : "indexId=" + indexId + ", catalogVersion=" + catalogVersion;
 
-    private long catalogActivationTimestampBusy(int catalogVersion) {
-        Catalog catalog = catalogService.catalog(catalogVersion);
-
-        assert catalog != null : catalogVersion;
-
-        return catalog.time();
-    }
-
-    // TODO: IGNITE-21771 Deal with catalog compaction
-    private void recoverStructuresBusy() {
-        int earliestCatalogVersion = catalogService.earliestCatalogVersion();
-        int latestCatalogVersion = catalogService.latestCatalogVersion();
-        int lwmCatalogVersion = catalogService.activeCatalogVersion(hybridTimestampToLong(lowWatermark.getLowWatermark()));
-
-        var tableVersionByIndexId = new HashMap<Integer, Integer>();
-        var readOnlyIndexes = new HashSet<ReadOnlyIndexInfo>();
-
-        var stoppingActivationTsByIndexId = new HashMap<Integer, Long>();
-        var previousCatalogVersionIndexIds = Set.<Integer>of();
-
-        for (int catalogVersion = earliestCatalogVersion; catalogVersion <= latestCatalogVersion; catalogVersion++) {
-            int finalCatalogVersion = catalogVersion;
-
-            var indexIds = new HashSet<Integer>();
-
-            catalogService.indexes(finalCatalogVersion).forEach(index -> {
-                tableVersionByIndexId.computeIfAbsent(index.id(), i -> tableVersionBusy(index, finalCatalogVersion));
-
-                if (index.status() == STOPPING) {
-                    stoppingActivationTsByIndexId.computeIfAbsent(index.id(), i -> catalogActivationTimestampBusy(finalCatalogVersion));
+                if (indexMeta.status() == MetaIndexStatus.READ_ONLY) {
```

Comment:
I think we could move checking READ_ONLY status to the top, it looks like the simplest condition to check and the most important

Reply:
We cannot, because it is in this area that we are protected from changing the low watermark and we are guaranteed not to lose the `IndexMeta`.